components/engine_traits/src/memory_engine.rs

```diff
@@ -1,19 +1,110 @@
 // Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
 
-use std::fmt::Debug;
+use std::{cmp, fmt::Debug};
+
+use keys::{enc_end_key, enc_start_key};
+use kvproto::metapb;
 
 use crate::{Iterable, Snapshot, WriteBatchExt};
 
-/// RegionCacheEngine works as a region cache caching some regions (in Memory or
+/// RangeCacheEngine works as a range cache caching some ranges (in Memory or
 /// NVME for instance) to improve the read performance.
-pub trait RegionCacheEngine:
+pub trait RangeCacheEngine:
     WriteBatchExt + Iterable + Debug + Clone + Unpin + Send + Sync + 'static
 {
     type Snapshot: Snapshot;
 
-    // If None is returned, the RegionCacheEngine is currently not readable for this
+    // If None is returned, the RangeCaheEngine is currently not readable for this
     // region or read_ts.
-    // Sequence number is shared between RegionCacheEngine and disk KvEnigne to
+    // Sequence number is shared between RangeCaheEngine and disk KvEnigne to
     // provide atomic write
-    fn snapshot(&self, region_id: u64, read_ts: u64, seq_num: u64) -> Option<Self::Snapshot>;
+    fn snapshot(&self, range: CacheRange, read_ts: u64, seq_num: u64) -> Option<Self::Snapshot>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct CacheRange {
+    pub start: Vec<u8>,
+    pub end: Vec<u8>,
+}
+
+impl CacheRange {
+    pub fn new(start: Vec<u8>, end: Vec<u8>) -> Self {
+        Self { start, end }
+    }
+
+    pub fn from_region(region: &metapb::Region) -> Self {
+        Self {
+            start: enc_start_key(region),
+            end: enc_end_key(region),
+        }
+    }
+}
+
+impl PartialOrd for CacheRange {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        if self.end <= other.start {
+            return Some(cmp::Ordering::Less);
+        }
+
+        if other.end <= self.start {
+            return Some(cmp::Ordering::Greater);
+        }
+
+        if self == other {
+            return Some(cmp::Ordering::Equal);
+        }
+
+        None
+    }
+}
+
+impl Ord for CacheRange {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        let c = self.start.cmp(&other.start);
+        if !c.is_eq() {
+            return c;
+        }
+        self.end.cmp(&other.end)
+    }
+}
+
+impl CacheRange {
+    // todo: need to consider ""?
+    pub fn contains(&self, other: &CacheRange) -> bool {
+        self.start <= other.start && self.end >= other.end
+    }
+
+    pub fn overlaps(&self, other: &CacheRange) -> bool {
+        self.start < other.end && other.start < self.end
+    }
+
+    pub fn is_sibling(&self, other: &CacheRange) -> bool {
+        self.start == other.end || self.end == other.start
+    }
+
+    pub fn split_off(&self, key: &CacheRange) -> (CacheRange, CacheRange) {
```

Comment:
Why the `key` is CacheRange instead of a single key?

Reply:
Because we are evicting a range.

Reply:
Say, now we have a range k01 - k10, and we are to evict k03-k05. After the eviction, there should be two ranges: k01-k03, and k05-k10.